File: haskell/day 1.hs
data Change = Increase | Decrease
  deriving (Show)

type Window = [Int]

isIncrease :: Change -> Bool
isIncrease Increase = True
isIncrease _ = False

sampleData :: [Integer]
sampleData = [199, 200, 208, 210, 200, 207, 240, 269, 260, 263]

transform :: [a] -> [(a, a)]
transform [] = []
transform [x] = []
transform (x : y : rest) = (x, y) : transform (y : rest)

toWindows :: [Int] -> [Window]
toWindows [] = []
toWindows [x] = []
toWindows [x, y] = []
toWindows (x : y : z : rest) = [x, y, z] : toWindows (y : z : rest)

getChange :: Int -> Int -> Change
getChange x y =
  if y > x
    then Increase
    else Decrease

countIncreases :: [Int] -> Int
countIncreases d = length $ filter isIncrease $ map (uncurry getChange) $ transform d

main :: IO ()
main = do
  input <- lines <$> readFile "input.txt"
  let inputData = map read input :: [Int]
  let partOne = countIncreases inputData
  let partTwo = countIncreases $ map sum $ toWindows inputData
  print partOne
  print partTwo

File: 2021/kotlin/src/main/kotlin/day 1.kt
private val sampleData = listOf(199, 200, 208, 210, 200, 207, 240, 269, 260, 263)

private enum class Change {
    Increase,
    Decrease
}

private typealias Window = List<Int>

private fun transform(vals: List<Int>): List<Pair<Int, Int>> {
    return when (vals.size) {
        0, 1 -> emptyList()
        else -> listOf(Pair(vals[0], vals[1])) + transform(vals.slice(1 until vals.size))
    }
}

private fun getChange(p: Pair<Int, Int>): Change {
    val x = p.first
    val y = p.second
    return if (y > x) Change.Increase else Change.Decrease
}

private fun countIncreases(vals: List<Int>): Int {
    return transform(vals).map(::getChange).filter { it == Change.Increase }.size
}

private fun toWindows(vals: List<Int>): List<Window> {
    if (vals.size in 0..2)
        return emptyList()

    val window: Window = listOf(vals[0], vals[1], vals[2])
    return listOf(window) + toWindows(vals.slice(1 until vals.size))
}

fun main() {
    val inputData = Utils
        .getInputLines(1)
        .map(String::toInt)

    val partOne = countIncreases(inputData)
    val partTwo = toWindows(inputData)
        .map(Iterable<Int>::sum)
        .let(::countIncreases)

    println(partOne)
    println(partTwo)
}